Shorten the time spent checking instruction length

qwe661234 opened the pull request:
I think we can pass the value of instruction length in IR to the exception handler rather than verifying it each time the emulate function is called, because we only need the value of rv->compressed in the exception handler.

jserv commented:
Once we complete the instruction decoding, we can definitely know the length of instruction(s) pointed by specific address. Would it make sense?

qwe661234 commented:
> Once we complete the instruction decoding, we can definitely know the length of instruction(s) pointed by specific address. Would it make sense?

Yes, we can know the length of instruction once instruction decoding completes. However, we should send a value to the exception handler if we want to reduce the number of times we check the instruction length.

jserv commented:
> Yes, we can know the length of instruction once instruction decoding completes. However, we should send a value to the exception handler if we want to reduce the number of times we check the instruction length.

If we can look up the instruction length (RV32 vs. RV32C) from a specific address, there is no need to preserve extra parameter in exception handlers. Let's keep it simple and stupid (KISS principle).

qwe661234 commented:
> > Yes, we can know the length of instruction once instruction decoding completes. However, we should send a value to the exception handler if we want to reduce the number of times we check the instruction length.
> 
> If we can look up the instruction length (RV32 vs. RV32C) from a specific address, there is no need to preserve extra parameter in exception handlers. Let's keep it simple and stupid (KISS principle).

Sorry, I miss the important information that we get instruction length by parsing instruction opcode, instead of specific address. If we don't want to pass more parameter to exception handlers, we need to save more information in struct `riscv_t`.

jserv commented:
Since preliminary basic block was introduced, speculated execution can take place without frequent instruction length checks. That is, the explicit transition from 32-bit ISA to 16-bit counterpart (vice versa) can be lazily evaluated in a basic block.

qwe661234 commented:
> Since preliminary basic block was introduced, speculated execution can take place without frequent instruction length checks. That is, the explicit transition from 32-bit ISA to 16-bit counterpart (vice versa) can be lazily evaluated in a basic block.

I think the idea above is checking instruction length in a basic block. However, there is a situation I am unsure whether it will occur or not. If a 32-bit instruction and a 16-bit compressed instruction coexist in the same basic block, it is unsafe to determine whether the instruction is compressed in basic block.

Risheng1128 commented:
We need to know the instruction is compressed or not when the exception occurred. So, I think that `rv->compressed` can be determined until the exception occurred.

For example, rewrite instruction `jal` as following:
```diff
case rv_insn_jal: { /* JAL: Jump and Link */
        ...
        /* check instruction misaligned */
        if (insn_is_misaligned(rv->PC)) {
+           rv->compressed = false;
            rv_except_insn_misaligned(rv, pc);
            return false;
        }
        /* can branch */
        return true;
    }
```
Therefore, we don't need to pass instruction length to exception handler and avoid to check `rv->compressed` frequently.

Risheng1128 commented on src/emulate.c:
```diff
@@ -254,10 +256,14 @@ static bool insn_is_misaligned(uint32_t pc)
 {
     /* check instruction is compressed or not */
-    rv->compressed = (ir->insn_len == INSN_16);
+    // rv->compressed = (ir->insn_len == INSN_16);
```
This comment should be removed.

Risheng1128 commented on src/emulate.c:
```diff
@@ -254,10 +256,14 @@ static bool insn_is_misaligned(uint32_t pc)
 }
 
+#define RV_Op(OPCODE) static bool op_##OPCODE(riscv_t *rv, rv_insn_t *ir)
+#define EMU_Op(OPCODE, RV, IR) op_##OPCODE(RV, IR)
```
I don't find the other code call these macros. Maybe, I think it is redundant.

qwe661234 commented:
> We need to know the instruction is compressed or not when the exception occurred. So, I think that `rv->compressed` can be determined until the exception occurred.
> 
> For example, rewrite instruction `jal` as following:
> 
> ```diff
> case rv_insn_jal: { /* JAL: Jump and Link */
>         ...
>         /* check instruction misaligned */
>         if (insn_is_misaligned(rv->PC)) {
> +           rv->compressed = false;
>             rv_except_insn_misaligned(rv, pc);
>             return false;
>         }
>         /* can branch */
>         return true;
>     }
> ```
> 
> Therefore, we don't need to pass instruction length to exception handler and avoid to check `rv->compressed` frequently.

Ok, I think it is better idea, I will follow this idea.

Risheng1128 commented on src/emulate.c:
```diff
@@ -290,6 +287,7 @@ static bool emulate(riscv_t *rv, rv_insn_t *ir)
             rv->X[ir->rd] = pc + ir->insn_len;
         /* check instruction misaligned */
         if (insn_is_misaligned(rv->PC)) {
+            rv->compressed = (ir->insn_len == INSN_16);
```
It should be `rv->compressed = false;` since instruction `jal` is always uncompressed. Therefore, we just assign an exact value.

qwe661234 replied:
Done.

Risheng1128 commented on src/emulate.c:
```diff
@@ -312,6 +310,7 @@ static bool emulate(riscv_t *rv, rv_insn_t *ir)
             rv->X[ir->rd] = pc + ir->insn_len;
         /* check instruction misaligned */
         if (insn_is_misaligned(rv->PC)) {
+            rv->compressed = (ir->insn_len == INSN_16);
```
Ditto, rewite the code as `rv->compressed = false;`.

jserv commented on src/riscv.h:
```diff
@@ -144,10 +144,10 @@ riscv_word_t rv_get_reg(riscv_t *rv, uint32_t reg);
 
 /* environment call handler */
-void ecall_handler(riscv_t *rv);
+void ecall_handler(riscv_t *rv, uint8_t insn_len);
```
Avoid changing the function prototype.

Risheng1128 commented on src/emulate.c:
```diff
@@ -1336,12 +1354,14 @@ void rv_step(riscv_t *rv, int32_t cycles)
 void ebreak_handler(riscv_t *rv)
 {
     assert(rv);
+    rv->compressed = false;
```
In ebreak case, there are two cases — `ebreak` and `c.ebreak`. Both of them execute function `ebreak_handler`, but their instruction length is different. Assign the value to `rv->compressed` in ebreak case respectively.

Risheng1128 commented on src/emulate.c:
```diff
@@ -1336,12 +1354,14 @@ void rv_step(riscv_t *rv, int32_t cycles)
 void ecall_handler(riscv_t *rv)
 {
     assert(rv);
+    rv->compressed = false;
```
For the consistency, I think we can move it to the function `emulate`.

qwe661234 replied:
OK

Risheng1128 commented on src/emulate.c:
```diff
@@ -551,9 +551,11 @@ static bool emulate(riscv_t *rv, rv_insn_t *ir)
         rv->X[ir->rd] = rv->X[ir->rs1] & rv->X[ir->rs2];
         break;
     case rv_insn_ecall: /* ECALL: Environment Call */
+        rv->compressed = (ir->insn_len == INSN_16);
```
Rewrite as `rv->compressed = false;` since `ecall` is uncompressed instruction.

Risheng1128 commented on src/emulate.c:
```diff
@@ -551,9 +551,11 @@ static bool emulate(riscv_t *rv, rv_insn_t *ir)
         rv->io.on_ecall(rv);
         return true;
     case rv_insn_ebreak: /* EBREAK: Environment Break */
+        rv->compressed = (ir->insn_len == INSN_16);
```
It should be `rv->compressed = false;`. In the other hand, `c.ebreak` case should add `rv->compressed = true;`

Risheng1128 commented:
The commit `Move instruction length checking` should merge into the commit `Reduce instruction length checking`.

jserv reviewed (changes requested):
Read https://cbea.ms/git-commit/ carefully for writing informative git commit messages.
You should explain the purpose of the proposed changes as well as their advantages.

jserv reviewed (changes requested):
Again, read https://cbea.ms/git-commit/ carefully.
* [Wrap the body at 72 characters](https://cbea.ms/git-commit/#wrap-72)
* [Use the body to explain what and why vs. how](https://cbea.ms/git-commit/#why-not-how)

Last but not least, clearly state how the proposed changes would be beneficial.

jserv reviewed (changes requested):
Do you understand what [Wrap the body at 72 characters](https://cbea.ms/git-commit/#wrap-72) means?

jserv commented:
By the way, use [QuillBot](https://quillbot.com/) to improve your writing.

qwe661234 commented:
> Do you understand what [Wrap the body at 72 characters](https://cbea.ms/git-commit/#wrap-72) means?

I see, I need to change line when characters in a line more than 72.

qwe661234 committed:
Shorten the time spent checking instruction length

To avoid frequent instruction length checking, move checking into
exception handler inoking block.

Previously, we had to check the instruction length each time the
function 'emulate' was called. Actually, we only need to check it when
the exception occurs; otherwise, we can simply set 'rv->compressed' to
the appropriate value before running the exception handler.

By doing so, we can avoid the overhead of checking instruction length if
an exception does not happen.

Resolves: #92
